### ISSUE TITLE
[HOLD for payment 2023-06-29] [$1000] If you press and hold on a message and then release the press, the hover effect will be lost 

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#16519]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open the app
2. Open any chat
3. Press and hold on message sender name
4. Release it the press below the name

## Expected Result:
The message hover effect should not be lost: the menu (of emojis, copy / mark unread / edit) should be displayed 

## Actual Result:
The message hover effect is lost: the menu (of emojis, copy / mark unread / edit) stops showing up

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.89-0
**Reproducible in staging?:** needs reproduction
**Reproducible in production?:** needs reproduction
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/227734378-0c046c5b-321d-4b97-9fd7-f1f8cad4869d.mov

https://user-images.githubusercontent.com/62073721/229502257-ac473305-19a0-45d4-bc7e-14bba4b47534.mov

**Expensify/Expensify Issue URL:**
**Issue reported by:** @getusha 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1679753778843439

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01d02ea979e978409c</li>
        <li>Upwork Job ID: 1644282540258504704</li>
        <li>Last Price Increase: 2023-06-06</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @laurenreidexpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by MelvinBot** [comment]
@laurenreidexpensify Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by MelvinBot** [comment]
@laurenreidexpensify Eep! 4 days overdue now. Issues have feelings too...

**Comment by laurenreidexpensify** [comment]
Apols for delay in reviewing this. Asking for a better video in the thread - I'm struggling to follow repro steps / identify the bug based on current evidence. Thanks

**Comment by laurenreidexpensify** [comment]
@Beamanator confirmed he can reproduce, have updated the O/P with a touch more detail

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01d02ea979e978409c

**Comment by MelvinBot** [comment]
Triggered auto assignment to @MitchExpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @sobitneupane (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @puneetlath (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by tienifr** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The message hover effect is lost after long pressing on the author's name, then release and click below the name.

### What is the root cause of that problem?
1. When we long press the author's name, the name component becomes focused
2. When we release the long press, the author's name will still have the focused since we didn't yet focus on any other element.
3. When we click on the space below the name, the `PressableWithSecondaryInteraction` here https://github.com/Expensify/App/blob/64f4a7814e87bd7c4d085776abf5cb8a67b9e8bc/src/pages/home/report/ReportActionItem.js#L241 will get the focus since it's the first focusable element that wraps the clicked point (it has `tabindex` defined).
4. The above component is focused, making the author's name component blurred
5. The blur (actually it's `focusout` in React) will bubble to the parent elements of the author's name, which include the `Hoverable` here https://github.com/Expensify/App/blob/64f4a7814e87bd7c4d085776abf5cb8a67b9e8bc/src/pages/home/report/ReportActionItem.js#L250
6. The `onBlur` of `Hoverable` is triggered, we have the safe check here https://github.com/Expensify/App/blob/f133eb4b536cd0fc8d7476720ae3cb3785ba6beb/src/components/Hoverable/index.js#L107 which will not set the `hovered` to false if the element receiving focus is a child of `Hoverable`. This safe check, however, will not work since the element receiving focus is `PressableWithSecondaryInteraction`, which is the direct parent of `Hoverable`.

We can see the root cause is the `6` here, the message container is actually receiving focus and should have been caught by the safe-check, but it fails since the `Hoverable` is not the parent of that message container.

### What changes do you think we should make in order to solve the problem?

The simple fix is to reorder the `Hoverable` and `PressableWithSecondaryInteraction`, a.k.a make the `Hoverable` the parent element of `PressableWithSecondaryInteraction` instead of the other way around. Since they are both wrapper elements which add functionality (hover, secondary interaction) that have no style, it will not cause any regressions.

After doing that, in step `6` above the hovered state will not be lost since the safe-check will now understand that the element receiving focus is a child of `Hoverable`.

There's another issue I found while investigating, the `InlineSystemMessage` here https://github.com/Expensify/App/blob/64f4a7814e87bd7c4d085776abf5cb8a67b9e8bc/src/pages/home/report/ReportActionItem.js#L306 is currently part of the `Pressable` but not the `Hoverable`, which means it will not trigger hovering of the whole report action item if we hover on it, but it will react to the long press as a part of the report action item.

I assume here that we want it not to be part of the report action item, we should move it out of the `Hoverable` as part of the above change. Or if it's part of the report action item we can keep it inside `PressableWithSecondaryInteraction`. With this there'll be consistency when long press/hover on that system message.

### What alternative solutions did you explore? (Optional)
NA

**Comment by moezh** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The hover effect on a message is lost if we press and hold the message sender's name or avatar and then release it on the space below (e.g on message).

### What is the root cause of that problem?
Pressing and holding on the sender's name or avatar focuses on it, then clicking on the space below it (e.g on message) focuses on the first focusable element `PressableWithSecondaryInteraction`.

### What changes do you think we should make in order to solve the problem?
We can prevent default on mouse down event so focus doesn't change by adding `onMouseDown={e => e.preventDefault()}` to the `PressableWithSecondaryInteraction` (in `ReportActionItem.js`).

### What alternative solutions did you explore? (Optional)
none

**Comment by tienifr** [comment]
@getusha I think the alternative solution I mentioned there is different from your suggested approach. I don't understand much how adding the `stopPropagation` inside `Hoverable` as you suggested will fix the issue, maybe I'm just missing something 😃.

**Comment by tienifr** [comment]
@getusha out of curiosity, wouldn't it still call `this.setIsHovered(false);` even after you add the `el.stopPropagation` in that block? just wondering how it works 😁

**Comment by sobitneupane** [comment]
@getusha Can you please add some explanation in "What is the root cause of that problem?" section in your proposal. Please do include flow of `onBlur` event and how is it interfering with the user's focus on the message.

**Comment by sobitneupane** [comment]
@getusha If you are still interested, bump on [this](https://github.com/Expensify/App/issues/16519#issuecomment-1501605876).

**Comment by laurenreidexpensify** [comment]
Note to myself: Offer sent to Sobit (C+) + Getusha (bug report)

**Comment by sobitneupane** [comment]
@parasharrajat Yup. I can confirm that reverting https://github.com/Expensify/App/pull/14377 fixes this issue.

**Comment by laurenreidexpensify** [comment]
@getusha can you link to the changes you're referring to? Thanks

**Comment by sobitneupane** [comment]
Looks like @parasharrajat is refactoring the `Hoverable` component in https://github.com/Expensify/App/pull/17422

**Comment by moezh** [comment]
We can prevent default on mouse down event to occur in the first place, so focus doesn't change.
Please check: https://github.com/Expensify/App/issues/16519#issuecomment-1500257474

**Comment by puneetlath** [comment]
Still on hold.

**Comment by laurenreidexpensify** [comment]
Still held on 

- #16052

**Comment by laurenreidexpensify** [comment]
@puneetlath @sobitneupane looks like #16052 was deployed to prod last week - we good to get this off hold now right?

**Comment by sobitneupane** [comment]
Testing if the issue still exists: Yes, It is still reproducible.

Yes, we are good to get this off hold @laurenreidexpensify.

**Comment by laurenreidexpensify** [comment]
@sobitneupane any thoughts ^^ ?

**Comment by laurenreidexpensify** [comment]
@arielgreen I am OOO til 19 June adding you to help keep moving along til I'm back. Thanks

**Comment by arielgreen** [comment]
@sobitneupane please review proposals asap

**Comment by sobitneupane** [comment]
Thanks for the proposal everyone.

[Third proposal](https://github.com/Expensify/App/issues/16519#issuecomment-1500172374) from @getusha makes sense and looks good to me.

🎀👀🎀 C+ reviewed

cc: @puneetlath

**Comment by puneetlath** [comment]
@arielgreen @laurenreidexpensify FYI, I'll be OOO next week. If the PR needs review before I'm back. Please remove my assignment and re-add the engineering label to get a new internal reviewer. Thanks!

**Comment by sobitneupane** [comment]
@arielgreen Since Puneet is OOO, can we assign new CME to get the PR merged.

**Comment by sobitneupane** [comment]
Since Puneet is OOO, @joelbettner https://github.com/Expensify/App/pull/20561 needs CME review.

**Comment by arielgreen** [comment]
Reassigning BZ, team change

**Comment by twisterdotcom** [comment]
Assigned @joelbettner to the PR.

**Comment by laurenreidexpensify** [comment]
@sobitneupane can you please complete steps above and confirm re: regression details thanks

**Comment by laurenreidexpensify** [comment]
@sobitneupane @getusha I had to send new offers in Upwork cos the old ones expired! If you accept will arrange payment thanks

**Comment by sobitneupane** [comment]
> **BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:
> 
> * [ ]  [@sobitneupane] The PR that introduced the bug has been identified. Link to the PR:

https://github.com/Expensify/App/pull/14377
> * [ ]  [@sobitneupane] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

https://github.com/Expensify/App/pull/14377#issuecomment-1617584024

> * [ ]  [@sobitneupane] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

This is a edge case and I don't think this could have been caught earlier.

**Comment by sobitneupane** [comment]
This issue is an edge case. I am not sure if we should add regression or not. 

**Regression Test Proposal:**

1. Open any chat
2. Press and hold on message sender name
3. Release it the press below the name (on the blank side)
4. Verify that the hover effect is not lost

DO we agree 👍 or 👎

**Comment by laurenreidexpensify** [comment]
Thanks @sobitneupane - can you accept job in Upwork so we can issue payment and close

@getusha can confirm payment has been issued 👍🏽

**Comment by laurenreidexpensify** [comment]
@getusha done! sorry for not catching that sooner. 

Waiting to issue payment to @sobitneupane when he accepts in Upwork and then we can close this one.

**Comment by sobitneupane** [comment]
@laurenreidexpensify Accepted the offer.
